aspnetcore/migration/20_21.md

```diff
@@ -0,0 +1,123 @@
+* Provides an overview of the changes to the ASP.NET Core web application templates.
+
+A quick way to get an overview of the changes in 2.1 is to:
+
+* Create an ASP.NET Core 2.0 web app named WebApp1.
+* Commit the WebApp1 in a source control system.
+* Delete WebApp1 and create an ASP.NET Core 2.1 web app named WebApp1 in the same place.
+* Review the changes in the 2.1 version.
+
+This article provides an overview on migration to  ASP.NET Core 2.1. It does not contain a complete list of all changes needed to migrate to version 2.1. Some projects might require more steps depending on the options selected when the project was created and modifications made to the project.
+
+## Update the project file to use 2.1 versions
+
+Update the *.csproj* project file:
+
+* Change `<TargetFramework>netcoreapp2.0</TargetFramework>` to the 2.1 version, that is `<TargetFramework>netcoreapp2.1</TargetFramework>`.
+* Replace the version specified "Microsoft.AspNetCore.All"  package reference with the versionless "Microsoft.AspNetCore.App" package reference. See [Migrating from Microsoft.AspNetCore.All to Microsoft.AspNetCore.App](xref:fundamentals/metapackage#migrate) and [Microsoft.AspNetCore.App metapackage](xref:fundamentals/metapackage-app).
+* Remove all references to `<DotNetCliToolReference>` elements for "Microsoft.AspNetCore", "Microsoft.VisualStudio", and "Microsoft.EntityFrameworkCore" packages. These tools have been replaced by global tools.
+* Remove the `<DotNetCliToolReference>` elements for any "Microsoft.AspNetCore" packages. These have been replaced by global tools.
+
+The following markup shows the template generated.csproj* project file:
+
+[!code-xml[Main](20_21/sample/WebApp1.csproj)]
+
+## Changes to take advantage of the new code-based idioms that are recommended in ASP.NET Core 2.1
+
+### Changes to Main
+
+The following images show the changes made to the templated generated *Program.cs* file.
+
+![old version differences](20_21/_static/main20.png)
+
+The preceding image shows the 2.0 version with the replacements in red.
+
+The following image shows the 2.1 code. The code in green replaced the 2.0 version:
+
+![new version differences](20_21/_static/main21.png)
+
+The following code shows the 2.1 version of *Program.cs*:
+
+[!code-csharp[Main](20_21/sample/Program.cs?name=snippet)]
+
+The new `Main` replaces the call to `BuildWebHost` with [CreateWebHostBuilder](/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1.createwebhostbuilder?view=aspnetcore-2.1#Microsoft_AspNetCore_Mvc_Testing_WebApplicationFactory_1_CreateWebHostBuilder). [WebHostBuilder](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.hosting.webhostbuilder?view=aspnetcore-2.0) was introduced to make it easier to write integration tests. For more information, see [Integration tests](xref:test/integration-tests) and  [Razor Pages unit and integration tests](xref:test/razor-pages-tests).
```

Comment:
U can take the `CreateWebHostBuilder` API link down to ...

> \[CreateWebHostBuilder](/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1.createwebhostbuilder)

Instead of the concrete `WebHostBuilder`, describe the interface ...

> \[IWebHostBuilder](/dotnet/api/microsoft.aspnetcore.hosting.iwebhostbuilder)

I view the introduction of `IWebHostBuilder` for integration tests as an esoteric implementation detail, so I don't surface the reason for its existence or its mechanics in the topic. That may change on review, but I hope not given the length and complexity of the topic. Therefore, I recommend (at least today) that you strip the last two sentences and remove the cross-link that to the testing topics. If you want to keep it tho, I recommend it go like this ...

> [IWebHostBuilder](/dotnet/api/microsoft.aspnetcore.hosting.iwebhostbuilder) is introduced to support new [integration test infrastructure](xref:test/integration-tests).

... that way it doesn't suggest that they'll get further explanation on it in the test topic.